docs/proposals/memory-management.md

```diff
@@ -0,0 +1,165 @@
+
+   a. `Guaranteed`: Memory is always guaranteed up to the request. These containers will not OOM unless they exceed their request, or system processes consume too much memory, or the kernel overhead is high and the memory in the node is being fully consumed. 
+
+   b. `Burstable`: Memory is guaranteed up until the usage exceeds user specified `Request`. These containers do not get killed as soon as they exceed their request. Kernel memory overhead and system processes memory usage will impact memory availability for this class of containers. They are more likely to be killed by the kernel OOM killer once they exceed their request. 
+
+   c. `Best-Effort`: Memory is never guaranteed. This class of containers are the first ones to be killed by the kernel whenever there is memory pressure on the node. 
+
+6. Detect OOM kills on containers: Docker tracks OOM kills on its containers. Kubelet exposes this information via [ContainerStatus.State](../pkg/api/v1/types.go#L1033), where the state is set to `ContainerStateTerminated` and `Reason` is set to `OOM Killed`.
+
+#### What are some of the known limitations?
+
+1. The scheduler does not take into account the total usage on the node. Hence, memory can be easily over-committed. Whenever a job is scheduled without resource requests, the node becomes over-committed. 
+
+2. Swapping is turned off by default. Kubelet does not track or manage swap files or partitions.
+
+3. A container in a pod with restart policy set to `Always` will be restarted continuously even if it OOMs. 
+
+4. Other than the Guaranteed class, containers in other classes don't have a memory limit. A `Best-effort` container can use a lot of memory and result in high tail latencies for `Guaranteed` and `Burstable` class containers.
+
+5. Kubelet does not act on OOM kills. A container can choose to continue running when one or more processes in the container were killed by the kernel OOM killer. 
+
+6. In the event of an OOM kill due to system memory pressure, the kubelet cannot identify the container of the killed process.
+
+7. Memory usage of system daemons is unrestricted. This results in varying amount of memory availability on the node.
+
+8. All burstable containers that exceed their request are equally likely to be killed. Ideally, the job that exceeds the most should be killed. 
+  
+##### How to address these limitations?
+
+We can improve memory management by making use of [kernel memcg features](https://www.kernel.org/doc/Documentation/cgroups/memory.txt), and by modifying the current system behavior. Some of these improvements are required in the short-term. The rest are either complex or require significant changes to the system.
+
+###### What can be done right away?
+
+1. Cleanup the entire container, whenever a OOM kill is detected in a container. Users should be able to override the default behavior for containers that have an in-built process manager that can handle OOM kills.
```

Comment:
I can't tell if you are suggesting that am OOM kill results in a pod becoming terminated in all cases, or only certain cases... does it apply only to best effort, or also burstable, or also guaranteed?  Maybe you are just saying this is design option, and not proposing a specific solution at this point in the doc?